drl/nn/nn.py

```diff
@@ -0,0 +1,447 @@
+	'keep_prob': None  # Percentage of neurons to keep on.
+}
+
+"""
+Default options are set to the atari convnet used in:
+    'Human-level control through deep reinforcement learning, Volodymyr Mnih, et al.' - https://www.nature.com/nature/journal/v518/n7540/full/nature14236.html
+    'Playing Atari with Deep Reinforcement Learning, Volodymyr Mnih, et al.' - https://arxiv.org/pdf/1312.5602.pdf
+    Lua code: https://sites.google.com/a/deepmind.com/dqn/
+"""
+convnet_options = {
+	'image_size': 84,  # Square image dimensions.
+	'channels': 4,  # Number of image channels.
+
+	'n_conv': 3,  # Number of convolutional layers.
+	'conv_filters': [32, 64, 64],  # Number of filters in each convolutional layer.
+	'conv_kernel_sizes': [8, 4, 3],  # Kernel sizes for each of the convolutional layer.
+	'conv_strides': [4, 2, 1],  # Stride sizes for each of the convolutional layer.
+
+	'n_fc': 1,  # Number of fully-connected layers.
+	'fc_units': [512],  # Number of output units in each fully-connected layer.
+
+	'activation_type': 'relu',  # Activation function for all your layers.
+	'loss_type': 'mse',  # Loss function you would like to use.
+	'pooling': False,  # Switch Pooling on/off. By default it is off.
+	'pooling_type': 'max_pooling',  # Type of pooling you want to apply. Default is set to max pooling.
+	'pooling_layers': [],  # List containing layer numbers where you want pooling to be applied.
+	'pooling_ksize': 2,  # Kernel size to be used for the pooling operation.
+	'pooling_stride': 2,  # Stride size to be used for the pooling operation.
+	'batch_norm': False,  # Switch Batch Normalization on/off. By default it is off.
+	'dropout': False,  # Switch dropout on/off. By default it is off.
+	'dropout_layers': [],  # List containing layer numbers where you want dropout.
+	'keep_prob': None  # Percentage of neurons to keep on.
+}
+
+
+class NN(object):
+	def __init__(self, mode, n_actions, n_obs=None, network_type=None, network_name=None, options_in=None):
+		"""
+		Constructs 'NN' object.
+
+			:param mode: 'train' or 'test'
+			:param n_actions: number of actions
+			:param n_obs: number of observations
+			:param network_type: 'conv' or 'fc'
+			:param network_name: gives the user an option to name networks in order to avoid confusion in
+								 case of more than one network
+			:param options_in: used to change default nn_options/convnet_options
+		"""
+		self.mode = mode
+		if self.mode != 'train' and self.mode != 'test':
+			color_print("ERROR: Set mode as either 'train' or 'test!'", color='red', mode='bold')
+			exit()
+		self.n_actions = n_actions
+		self.n_obs = n_obs
+		self.network_type = network_type
+		self.network_name = network_name
+
+		if self.network_type == 'conv':
+			if options_in is not None:
+				convnet_options.update(options_in)
+			self.options = convnet_options
+		else:
+			if options_in is not None:
+				nn_options.update(options_in)
+			self.options = nn_options
+
+		if self.mode == 'test':
```

Comment:
How does this work in practice? The reinforcement learning cycle consists out of taking a step (inference/test) followed by updating parameters (train). I think it should have full access during a step and be in test mode during updating. 

I had a similar problem when implementing batch normalization, when you take a step you only have one data point, if you are still in train mode your single observation will become zero all the time. See in utilities/tfnetwork the bn_layer function for how i implemented this using a tf.placeholder.